### PR TITLE
ATF: Support secure connection

### PR DESCRIPTION
**Pull request by mghiumiusliu**
This proposal describes support of security sessions in ATF (Automated Test Framework). ATF should be able to test SDL security feature.

Main ATF features for checking SDL security feature:
```
Support of secure SDL sessions;
Support of testing broken handshake;
Support of sending raw data in secure channel;
Support of expected secure requests\responses\notifications;
Support Transport Layer Security (TLS) and Datagram Transport Layer Security (DTLS).
```

### CODE REVIEW

**Comment by mghiumiusliu** [comment]
Proposal is Ready for review

File: proposals/NNNN-atf_security_proposal.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+ - Support Transport Layer Security (TLS) and Datagram Transport Layer Security (DTLS). 
+
+## Motivation
+
+Motivation of this proposal is to create ability of coverage the secure sessions. Which is the most important not covered part of SDL functionality.
+
+Manual testing of security SDL feature is slow and expensive. Also there is a big probability of error during manual testing.
+Testing automatization of all SDL use cases is the best option to be sure that new code does not break SDL functionality.
+
+ATF has possibility to cover almost all SDL use cases with automatic testing. But there is still some functionality that remains uncovered my ATF:
```
Comment (theresalech) [inline proposals/NNNN-atf_security_proposal.md]: Please update to "ATF has possibility to cover almost all SDL use cases with automatic testing, but there is still some functionality that remains uncovered by ATF:"

File: proposals/NNNN-atf_security_proposal.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+ - session.SendEncryptedRPC;
+ - session.ExpectEncryptedResponse;
+ - session.ExpectEncryptedNotification;
+ - session.SendPacket; 
+ - session.ExpectPacket;
+ 
+[mobile_session.lua](https://github.com/smartdevicelink/sdl_atf/blob/master/modules/mobile_session.lua) and 
+[mobile_session_impl.lua](https://github.com/smartdevicelink/sdl_atf/blob/master/modules/mobile_session_impl.lua) should be extended with secure session interfaces. 
+
+Should be added new ini file option: `SecurityProtocol`\
```
Comment (theresalech) [inline proposals/NNNN-atf_security_proposal.md]: Please change "Should be added new..." to "Should add new..."

File: proposals/NNNN-atf_security_proposal.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+Should be added new command line option: `security_protocol`\
```
Comment (theresalech) [inline proposals/NNNN-atf_security_proposal.md]: Please change "Should be added new..." to "Should add new..."

File: proposals/NNNN-atf_security_proposal.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+In case if this option is missed, use TLS security by default.
```
Comment (theresalech) [inline proposals/NNNN-atf_security_proposal.md]: Please update to "If this option is missed, use TLS security by default."

File: proposals/NNNN-atf_security_proposal.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+### Detailed design
+
+#### New component: 
+Will be added new component : *SecurityManager*
```
Comment (theresalech) [inline proposals/NNNN-atf_security_proposal.md]: Please update to "Will add new component"

File: proposals/NNNN-atf_security_proposal.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+##### connection.StartSecureSession:
+###### Description
+ Start secure session :
+  1. Send StartSession with `encrypted flag = true`;
+  2. Perform TLS/DTLS handshake; 
+  3. Expect StartSessionAck.
+  
+  This is a blocking call. Will block current execution until session won't be established or failed.
+###### Arguments:
+  - protocol (if missed, used one from console or ini file options).
```
Comment (theresalech) [inline proposals/NNNN-atf_security_proposal.md]: I think "used one" should be changed to "use one"

File: proposals/NNNN-atf_security_proposal.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+## Potential downsides
+
+- session.ExpectPacket - very common interface, it might need more deep analysis of input data. Also present possibility to reduce efficiency of the whole ATF system. This interface also allows tester to check any incoming data from SDL.
+
+- session.SendPacket - very low level interface, it might require a lot of additional complicated logic for constructing bytes to send in script. This interface also allows tester to send any data.
+
+## Impact on existing code
+
+Should be impacted only ATF code.
```
Comment (theresalech) [inline proposals/NNNN-atf_security_proposal.md]: "Should only impact ATF code."

File: proposals/NNNN-atf_security_proposal.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+If during implementation will be found some blocker issues in SDL, they should be fixed.
```
Comment (theresalech) [inline proposals/NNNN-atf_security_proposal.md]: "If blocker issues in SDL Core are found during implementation, the should be fixed."

**Comment by mghiumiusliu** [comment]
@theresalech Requested changes were introduced.
Thank you for support.
Proposal is Ready for Review.

**Pull request merged**